hoomd/mpcd/SphereGeometry.h

```diff
@@ -0,0 +1,209 @@
+#include "hoomd/BoxDim.h"
+
+#include <cmath>
+
+#ifdef NVCC
+#define HOSTDEVICE __host__ __device__ inline
+#else
+#define HOSTDEVICE inline __attribute__((always_inline))
+#include <string>
+#endif // NVCC
+
+namespace mpcd
+{
+namespace detail
+{
+//! Spherical droplet geometry
+/*!
+ * TODO: write proper description of what is being done in the class below
+ */
+class __attribute__((visibility("default"))) SphereGeometry
+    {
+    public:
+        //! Constructor
+        /*!
+         * \param R confinement radius
+         * \param bc Boundary condition at the wall (slip or no-slip)
+         */
+        HOSTDEVICE SphereGeometry(Scalar R, boundary bc)
+            : m_R(R), m_bc(bc)
+            { }
+
+        //! Detect collision between the particle and the boundary
+        /*!
+         * \param pos Proposed particle position
+         * \param vel Proposed particle velocity
+         * \param dt Integration time remaining
+         *
+         * \returns True if a collision occurred, and false otherwise
+         *
+         * \post The particle position \a pos is moved to the point of reflection, the velocity \a vel is updated
+         *       according to the appropriate bounce back rule, and the integration time \a dt is decreased to the
+         *       amount of time remaining.
+         */
+        HOSTDEVICE bool detectCollision(Scalar3& pos, Scalar3& vel, Scalar& dt) const
+            {
+
+            /*
+             * Detect if particle has left the spherical confinement.
+             * comparison: 'r2 < R*R' used in the following calculations is equal to 0(FALSE) if the particle is outside
+             * a sphere of radius R, 1(TRUE) if inside.
+             *
+             * We intentionally use < rather than <= in the comparison so that spurious collisions are not detected
+             * when a particle is reset to the spherical boundary. A particle landing exactly on the spherical boundary
+             * can be reflected in the next streaming step.
+             * TODO: Verify if the comparison can impact the dynamics incorrectly since we aim to have a dynamic spherical confinement.
+             */
+            Scalar r2 = dot(pos,pos);
+            // exit immediately if no collision is found or particle is not moving normal (radial direction) to the wall
+            // (since no new collision could have occurred if there is no normal motion)
+            /*
+            r(t) = sqrt(x**2 + y**2 + z**2)
+            radial velocity, r_dot = (x*vel_x + y*vel_y + z*vel_z)/r(t)
+
+            define variable vr = dot(pos,vel) to check for radial motion, for computational efficiency.
+
+            To avoid division by zero error later on (line 103), we exit immediately. (no collision could have occurred
+            if the particle speed is equal to zero)
+            */
+            Scalar vr = dot(pos,vel);
+            Scalar v2 = dot(vel,vel);
+            if (r2 < R*R || vr == Scalar(0) || v2 == Scalar(0))
```

Comment:
I think this should be:
```suggestion
            if (r2 <= R*R || v2 == Scalar(0))
```
If you look at the SlitGeometry (https://github.com/glotzerlab/hoomd-blue/blob/b6cfe0bc20b859d38a154a6efb5762e2a1cb8a27/hoomd/mpcd/SlitGeometry.h#L93-L100), you will see that the condition to exit is when the particle is `not z > H and not z < -H`, which is the same as `<=`. I also don't think it makes sense to check on `dot(v,r)` here, since `r` is not the normal at the point the particle exited the sphere.